File: src/ios/Founditure/Config/AppConfig.swift
//
// AppConfig.swift
// Founditure
//
// HUMAN TASKS:
// 1. Verify environment configuration in build settings
// 2. Set up appropriate provisioning profiles for each environment
// 3. Configure cache storage permissions in entitlements
// 4. Update Info.plist with required privacy descriptions
// 5. Set up keychain access groups if needed

// Foundation framework - iOS 14.0+
import Foundation

// Internal dependencies
import AppConstants

/// Environment enumeration for different application environments
/// Requirement: System Architecture - Manages client-side configuration for distributed cloud-based infrastructure
enum Environment {
    case development
    case staging
    case production
}

/// Configuration structure for application caching
/// Requirement: Mobile Applications - Implements native iOS application configuration with offline-first architecture
struct CacheConfiguration {
    let expirationInterval: TimeInterval
    let maxCacheSize: Int
    let clearOnLogout: Bool
}

/// Core configuration class managing application-wide settings and environment configurations
/// Requirement: Security Architecture - Implements security configuration and environment-specific settings
final class AppConfig {
    // MARK: - Properties
    
    /// Singleton instance
    static let shared = AppConfig()
    
    /// Current environment setting
    private(set) var environment: Environment
    
    /// Application version
    private(set) var appVersion: String
    
    /// Build number
    private(set) var buildNumber: String
    
    /// Dictionary of enabled features
    private(set) var enabledFeatures: [String: Bool]
    
    /// Cache configuration settings
    private(set) var cacheConfig: CacheConfiguration
    
    // MARK: - Initialization
    
    private init() {
        // Initialize with default development environment
        self.environment = .development
        
        // Load app version and build number from bundle
        let bundle = Bundle.main
        self.appVersion = bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "1.0.0"
        self.buildNumber = bundle.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "1"
        
        // Initialize feature flags from AppConstants
        self.enabledFeatures = [
            "offlineMode": Features.offlineMode,
            "imageRecognition": Features.imageRecognition,
            "locationSharing": Features.locationSharing,
            "pushNotifications": Features.pushNotifications,
            "chatEncryption": Features.chatEncryption,
            "biometricAuth": Features.biometricAuth
        ]
        
        // Setup default cache configuration
        self.cacheConfig = CacheConfiguration(
            expirationInterval: TimeInterval(App.cacheExpirationDays * 24 * 60 * 60),
            maxCacheSize: App.maxImageSize * App.maxImageCount,
            clearOnLogout: true
        )
    }
    
    // MARK: - Configuration Methods
    
    /// Configures application settings based on current environment
    /// - Parameter environment: The target environment to configure
    func configure(environment: Environment) {
        self.environment = environment
        
        // Load environment-specific configurations
        switch environment {
        case .development:
            configureDevelopment()
        case .staging:
            configureStaging()
        case .production:
            configureProduction()
        }
        
        // Initialize feature flags based on environment
        initializeFeatureFlags()
        
        // Setup cache settings
        setupCacheConfiguration()
        
        // Configure logging level
        configureLogging()
    }
    
    /// Checks if a specific feature is enabled
    /// - Parameter featureKey: The key of the feature to check
    /// - Returns: Boolean indicating if the feature is enabled
    func isFeatureEnabled(_ featureKey: String) -> Bool {
        return enabledFeatures[featureKey] ?? false
    }
    
    /// Updates cache configuration settings
    /// - Parameter config: New cache configuration to apply
    func updateCacheConfig(_ config: CacheConfiguration) {
        // Validate cache configuration
        guard config.expirationInterval > 0 && config.maxCacheSize > 0 else {
            return
        }
        
        // Update cache settings
        self.cacheConfig = config
        
        // Clear invalid cache if needed
        if config.clearOnLogout {
            clearInvalidCache()
        }
    }
    
    // MARK: - Private Methods
    
    private func configureDevelopment() {
        enabledFeatures["debugMode"] = true
        enabledFeatures["testFeatures"] = true
    }
    
    private func configureStaging() {
        enabledFeatures["debugMode"] = true
        enabledFeatures["testFeatures"] = false
    }
    
    private func configureProduction() {
        enabledFeatures["debugMode"] = false
        enabledFeatures["testFeatures"] = false
    }
    
    private func initializeFeatureFlags() {
        // Update feature flags based on environment
        switch environment {
        case .development:
            // Enable all features in development
            enabledFeatures.keys.forEach { enabledFeatures[$0] = true }
        case .staging:
            // Enable core features in staging
            enabledFeatures["offlineMode"] = true
            enabledFeatures["imageRecognition"] = true
        case .production:
            // Production features based on AppConstants
            enabledFeatures = [
                "offlineMode": Features.offlineMode,
                "imageRecognition": Features.imageRecognition,
                "locationSharing": Features.locationSharing,
                "pushNotifications": Features.pushNotifications,
                "chatEncryption": Features.chatEncryption,
                "biometricAuth": Features.biometricAuth
            ]
        }
    }
    
    private func setupCacheConfiguration() {
        let config = CacheConfiguration(
            expirationInterval: TimeInterval(App.cacheExpirationDays * 24 * 60 * 60),
            maxCacheSize: environment == .production ? App.maxImageSize * 2 : App.maxImageSize * App.maxImageCount,
            clearOnLogout: environment != .development
        )
        updateCacheConfig(config)
    }
    
    private func configureLogging() {
        // Set logging level based on environment
        switch environment {
        case .development:
            // Verbose logging for development
            print("Debug logging enabled")
        case .staging:
            // Warning and error logging for staging
            print("Warning logging enabled")
        case .production:
            // Error-only logging for production
            print("Error-only logging enabled")
        }
    }
    
    private func clearInvalidCache() {
        // Implementation for clearing invalid cache
        // This would typically involve clearing expired items and ensuring
        // cache size remains within configured limits
    }
}

File: src/android/app/src/main/kotlin/com/founditure/android/config/AppConfig.kt
/**
 * Human Tasks:
 * 1. Ensure BuildConfig.BUILD_TYPE is properly configured in build.gradle for each environment
 * 2. Configure proper Settings.Secure permissions in AndroidManifest.xml
 * 3. Verify environment-specific API endpoints in Constants.kt
 * 4. Set up proper security configurations for each environment
 * 5. Configure logging levels in build.gradle for different build variants
 */

package com.founditure.android.config

import android.provider.Settings
import com.founditure.android.BuildConfig
import com.founditure.android.util.API
import com.founditure.android.util.Database
import com.founditure.android.util.Security

/**
 * Primary configuration object for the Founditure Android application.
 * Implements requirements from:
 * - Mobile Applications (1.2 Scope/Core System Components/1. Mobile Applications)
 * - System Architecture (3.1 High-Level Architecture Overview)
 * - Security Architecture (3.6 Security Architecture)
 */

/**
 * Environment enumeration for different deployment environments
 * Implements requirements from System Architecture section 3.1
 */
enum class Environment {
    DEVELOPMENT,
    STAGING,
    PRODUCTION
}

/**
 * Log level enumeration for application logging
 * Implements requirements from System Architecture section 3.1
 */
enum class LogLevel {
    DEBUG,
    INFO,
    WARNING,
    ERROR
}

/**
 * Singleton object containing global application configuration
 * Implements requirements from Mobile Applications section 1.2 and Security Architecture section 3.6
 */
object AppConfig {
    // Build configuration properties
    val isProduction: Boolean = BuildConfig.BUILD_TYPE == "release"
    val isDebugMode: Boolean = BuildConfig.DEBUG
    val appVersion: String = BuildConfig.VERSION_NAME
    val appVersionCode: Int = BuildConfig.VERSION_CODE
    val deviceId: String = Settings.Secure.ANDROID_ID

    /**
     * Determines the current application environment based on build configuration
     * Implements requirements from System Architecture section 3.1
     */
    fun getCurrentEnvironment(): Environment {
        return when {
            isProduction -> Environment.PRODUCTION
            BuildConfig.BUILD_TYPE == "staging" -> Environment.STAGING
            else -> Environment.DEVELOPMENT
        }
    }

    /**
     * Returns the appropriate log level based on the current environment
     * Implements requirements from System Architecture section 3.1
     */
    fun getLogLevel(): LogLevel {
        return when (getCurrentEnvironment()) {
            Environment.DEVELOPMENT -> LogLevel.DEBUG
            Environment.STAGING -> LogLevel.INFO
            Environment.PRODUCTION -> LogLevel.WARNING
        }
    }

    /**
     * Checks if a specific feature flag is enabled for the current environment
     * Implements requirements from System Architecture section 3.1
     */
    fun isFeatureEnabled(featureKey: String): Boolean {
        // Feature flags configuration based on environment
        val featureFlags = mapOf(
            "offline_mode" to true,
            "real_time_chat" to (getCurrentEnvironment() != Environment.DEVELOPMENT),
            "ai_recognition" to (getCurrentEnvironment() != Environment.DEVELOPMENT),
            "location_fuzzing" to isProduction,
            "advanced_analytics" to isProduction
        )

        return featureFlags[featureKey] ?: false
    }

    /**
     * Get API configuration based on current environment
     * Implements requirements from System Architecture section 3.1
     */
    fun getApiConfig(): Map<String, String> {
        return mapOf(
            "baseUrl" to API.BASE_URL,
            "version" to API.API_VERSION,
            "timeout" to "${API.TIMEOUT_CONNECT}",
            "maxRetries" to "${API.MAX_RETRIES}"
        )
    }

    /**
     * Get database configuration based on current environment
     * Implements requirements from Mobile Applications section 1.2
     */
    fun getDatabaseConfig(): Map<String, Any> {
        return mapOf(
            "name" to Database.DATABASE_NAME,
            "version" to Database.DATABASE_VERSION,
            "maxCacheSize" to Database.MAX_CACHE_SIZE_MB,
            "cacheExpiry" to Database.CACHE_EXPIRY_MS
        )
    }

    /**
     * Get security configuration based on current environment
     * Implements requirements from Security Architecture section 3.6
     */
    fun getSecurityConfig(): Map<String, Any> {
        return mapOf(
            "algorithm" to Security.ENCRYPTION_ALGORITHM,
            "keySize" to Security.KEY_SIZE,
            "keystoreAlias" to Security.KEYSTORE_ALIAS,
            "tokenExpiry" to Security.TOKEN_EXPIRY
        )
    }
}